File: golang/1130. 叶值的最小代价生成树.go
func mctFromLeafValues(arr []int) int {
	res, stk := 0, []int{}
	for _, x := range arr {
		for len(stk) > 0 && stk[len(stk)-1] <= x {
			if len(stk) == 1 || stk[len(stk)-2] > x {
				res += stk[len(stk)-1] * x
			} else {
				res += stk[len(stk)-2] * stk[len(stk)-1]
			}
			stk = stk[:len(stk)-1]
		}
		stk = append(stk, x)
	}
	for len(stk) > 1 {
		res += stk[len(stk)-2] * stk[len(stk)-1]
		stk = stk[:len(stk)-1]
	}
	return res
}

File: 刷题/cpp/1130. 叶值的最小代价生成树.cpp
// 法一：动态规划
class Solution {
public:
    int mctFromLeafValues(vector<int>& arr) {
        int n=arr.size();
        vector<vector<int>> dp(n,vector<int>(n,INT_MAX/4)), mval(n,vector<int>(n));
        for(int j=0;j<n;j++){
            mval[j][j]=arr[j];
            dp[j][j]=0;
            for(int i=j-1;i>=0;i--){
                mval[i][j]=max(arr[i],mval[i+1][j]);
                for(int k=i;k<j;k++){
                    dp[i][j]=min(dp[i][j],dp[i][k]+dp[k+1][j]+mval[i][k]*mval[k+1][j]);
                }
            }
        }
        return dp[0][n-1];
    }
};

//法二：单调栈
class Solution {
public:
    int mctFromLeafValues(vector<int>& arr) {
        int res=0;
        stack<int> stk;
        for(int x:arr){
            while(!stk.empty()&&stk.top()<=x){
                int y=stk.top();
                stk.pop();
                if(stk.empty()||stk.top()>x){
                    res+=y*x;
                }else{
                    res+=stk.top()*y;
                }
            }
            stk.push(x);
        }
        while(stk.size()>=2){
            int x=stk.top();
            stk.pop();
            res+=stk.top()*x;
        }
        return res;
    }
};